ov2slam/ov2slam
# Updates supporting OV²SLAM usage in ROS2

# Changes

- ROS-related source code in `include/camera_visualizer.hpp`, `include/ros_visualizer.hpp`, `src/camera_visualizer.cpp`, `src/ov2slam.cpp`, and in `src/ov2slam_node.cpp` was changed to ROS2 version. Changes involved:
  - `#include <ros/ros.h>` -> `#include <rclcpp/rclcpp.hpp>`
  - updates in message header files include and in message creation
  - ROS node creation
  - ROS topics publishers and subscribers
  - TF transform broadcaster
  - getting ROS time for headers: `ros::Time(time)` -> `n_->get_clock()->now()`
  - `getNumSubscribers()` -> `get_subscription_count()`
- `package.xml` and `CMakeLists.txt` updated to ROS2-supported versions
- the package version (in `package.xml`) was changed to **2.0.0**
- *pcl_ros* was replaced by *pcl_conversions* due to lack of full support for ROS2. Changes were applied in `include/map_manager.hpp` and `src/map_manager.cpp`
- changed `CMakeLists.txt` in *obindex2* and *ibow_lcd*, so now libraries are installed system-wide like e.g. _ceres-solver_ or _Sophus_, with namespaces, respectively, *obindex2* and *lcdetector*
- `build_thirdparty.sh` - LC-related packages are now installed as system-wide libraries
- updated `FindTBB.cmake` in _ceres-solver_ due to missing file issue, according to [solution](https://github.com/yppah56/OpenFPGA/commit/8cdaceb4358c3590e473bf7d623c46ab932e2e6a)
- backward-cpp was replaced by [backward_ros](https://github.com/pal-robotics/backward_ros)
- added a `docker` directory with a `Dockerfile` and usage description in `docker/README.md`
- added ROS2 version of `ov2slam_visualization.rviz` for topics visualization
- `README.md` adjusted description for ROS2
- added a `.gitignore` file to filter out tracked files by version control

# How to verify
1. Build a docker container according to *docker/README.md*.
2. In the docker container source environment (`source install/setup.bash`), and run `ov2slam_node`. One should provide a config file from the config folder in the package directory: `./src/ov2slam/parameters_files/`.

```shell
ros2 run ov2slam ov2slam_node src/ov2slam/parameters_files/parameters_file.yaml
```

3. In other terminal run bag2 with command `ros2 bag play <PATH TO ROS2 BAG FILE>`.
4. In the third terminal monitor published VO pose, e.g. with `ros2 topic hz /vo_pose` or run rviz2 topics visualizer (if docker runs with X forwarding) `rviz2 -d ./src/ov2slam/ov2slam_visualization.rviz`

## Evaluation
For evaluation purposes the [evo](https://github.com/MichaelGrupp/evo) package was utilized, calculating ATE, according to the [author](https://github.com/MichaelGrupp/evo/issues/247#issuecomment-586283971), with the command:

```shell
evo_ape tum reference_file.txt ov2slam_results_file.txt --align --pose_relation trans_part
```

The evaluation was performed on the [EuRoC](https://projects.asl.ethz.ch/datasets/doku.php?id=kmavvisualinertialdatasets) dataset. Reference files were extracted from ROS2 bags using `evo_traj bag2 <BAG DIR> <TOPIC NAME> --save_as_tum`


### Evaluation Results

Metrics were calculated using the *ov2slam_full_traj_wlc_opt.txt* results file with timestamps associated with *ov2slam_traj.txt*.

<div align="center">

| **Bag** | **ATE rmse [m]** |
|:---:|:---:|
| MH 01 | 0.106435 |
| MH 02 | 0.058447 |
| MH 03 | 0.195186 |
| MH 04 | 0.243443 |
| MH 05 | 0.267268 |
| V1 01 | 0.088002 |
| V1 02 | 0.115117 |
| V1 03 | 0.241767 |
| V2 01 | 0.092051 |
| V2 02 | 0.183751 |
| V2 03 | 1.687249 |

</div>

<div align="center">

<img src="https://github.com/ov2slam/ov2slam/assets/34938498/7044be6d-92c4-462c-aefd-6fcb5bf1cf10" alt="mh_01_map" width="300">
<img src="https://github.com/ov2slam/ov2slam/assets/34938498/b492cccf-5073-4ece-8220-56cbde85c96b" alt="mh_02_map" width="300">
<img src="https://github.com/ov2slam/ov2slam/assets/34938498/e9e01d31-f872-43e0-8e74-0dbc7752bd28" alt="mh_03_map" width="300">
<img src="https://github.com/ov2slam/ov2slam/assets/34938498/4a05574a-74ea-4376-acd8-5108818dbca8" alt="mh_04_map" width="300">
<img src="https://github.com/ov2slam/ov2slam/assets/34938498/0679b235-47ba-4bf4-94b3-28215b23d136" alt="mh_05_map" width="300">

<p>Results trajectory compared to reference trajectory, respectively for MH 01, MH 02, MH 03, MH 04, and MH 05 from the EuRoC dataset.</p>

</div>


### ROS2 bags
- dataset (in ROS1 bag format):
  - [EuRoC](https://projects.asl.ethz.ch/datasets/doku.php?id=kmavvisualinertialdatasets)
- conversion from ROS1 bag to ROS2 bag:
  - e.g. using Rosbags Python package:
    - [installation](https://ternaris.gitlab.io/rosbags/index.html): `pip install rosbags`
    - [conversion](https://ternaris.gitlab.io/rosbags/topics/convert.html): `rosbags-convert ros1.bag --dst ./ros2_bag/`
